[JENKINS-55663] Extend the Jenkinsfile Runner user documentation (#96)
jenkinsci/jenkinsfile-runner: DOCKER.md

You can package Jenkinsfile Runner with a specific Jenkins image and turn that into a Docker image.
This way, you can ensure people are running Jenkinsfile in the specific Jenkins environment, for example one that's identical to your production environment.

## Building the Vanilla image

> This repository offers a Dockerfile to generate your docker image.

Reviewer: to generate a JFR docker image?

Edit the `plugins.txt` to include any plugins you'd like to install. See [the Jenkins Docker README](https://github.com/jenkinsci/docker#preinstalling-plugins) for more information.

Then, build the Jenkinsfile Runner image like this:

> ```bash

Reviewer: Is this for Github's formatting?

Author: That enables syntax highlighting. I've seen it was used in other parts of the documentation and I've unified its use

The optional `JENKINS_VERSION` specifies the version of Jenkins core.

> Notice that the master workspace is mapped to `/build`.

Reviewer 2: ```suggestion
Note that the master workspace is mapped to `/build` by default.
```

This directory can be exposed as a volume.

## Usage
Run the image by providing path to `Jenkinsfile` . For example,

```bash
docker run --rm -v $PWD/test:/workspace jenkinsfile-runner:my-production-jenkins
```

> In same cases you can be interested in passing extra options.

Reviewer 2: ```suggestion
In same cases you may be interested in passing extra options.
```

> It can be done passing the `JAVA_OPTS` environment variable.

Reviewer 2: ```suggestion
Use the `JAVA_OPTS` environment variable to do that.
```

```bash
docker run --rm -e JAVA_OPTS="-Xms 256m" -v $PWD/test:/workspace jenkinsfile-runner:my-production-jenkins
```

> Also is possible to pass arguments to the Jenkinsfile or to execute it outside a sandbox environment with the `-ns` and `-a` options.

Reviewer: "It is also possible" is better I think.

> It is also possible to pass arguments to the Jenkinsfile or to execute it outside a sandbox environment with the `-ns` and `-a` options.

Reviewer 2: This line is a bit confusing. The example below clarifies it tho.
I would also recommend to put a warning that non-sandbox mode may pose potential security risks, and hence it should be used carefully

Author: Done!

```bash
docker run --rm -e DEBUG=true -p 5005:5005 -v $PWD/test:/workspace jenkinsfile-runner:my-production-jenkins
```

## Custom War Packager

> There is an alternative to Vanilla Dockerfile that gives a huge versatility when it comes to generate the docker images.

Reviewer 2: The flow also generates an off-Docker executable though we have never tested it. Better to keep it as is

See the demo for a better understanding on [how to build Jenkinsfile Runner with Custom WAR Packager](demo/cwp).

> Once the docker image is generated you can execute it in the same way as a Vanilla image with the only exception of the master workspace.

Reviewer: Why is the workspace an exception? The vanilla image works the same way.

Author: - Entrypoint in Vanilla image:
```
ENTRYPOINT ["/app/bin/jenkinsfile-runner-launcher", \
            "-w", "/app/jenkins",\
            "-p", "/usr/share/jenkins/ref/plugins",\
            "-f", "/workspace", \
            "--runWorkspace", "/build"]
```
- Entrypoint in Dockerfile generated by CWP:
```
ENTRYPOINT ["/app/bin/jenkinsfile-runner", \
            "-w", "/app/jenkins",\
            "-p", "/usr/share/jenkins/ref/plugins",\
            "-f", "/workspace/Jenkinsfile"]
```
The `runWorkspace` is not set by default to `/build` and that's why there is an exception and the next sentence is needed:

> If you are interested in mapping a volume and having access to it, both the -v docker option and the --runworkspace options needs to be specified.

Reviewer: See https://github.com/jenkinsci/custom-war-packager/pull/71/files#diff-07df933f2a2d07a256ae752fa9ac363fR58

Reviewer 2: Yes, needs fix

The default directory is `/build` as well but it can be overridden using the `--runworkspace` Jenkinsfile Runner option.

> A special case worth mentioning is the joint use of Custom War Packager and [Dependabot](https://dependabot.com) to maintain your docker image up to date.

Reviewer 2: Reference https://github.com/jenkinsci/ci.jenkins.io-runner as a reference impl?

Author: Sure thing!
